spring-cloud-task-stream/src/main/java/org/springframework/cloud/task/batch/listener/BatchEventAutoConfiguration.java

```diff
@@ -0,0 +1,157 @@
+
+import org.springframework.batch.core.ChunkListener;
+import org.springframework.batch.core.ItemProcessListener;
+import org.springframework.batch.core.ItemReadListener;
+import org.springframework.batch.core.ItemWriteListener;
+import org.springframework.batch.core.Job;
+import org.springframework.batch.core.JobExecutionListener;
+import org.springframework.batch.core.SkipListener;
+import org.springframework.batch.core.StepExecutionListener;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.cloud.stream.annotation.EnableBinding;
+import org.springframework.cloud.stream.annotation.Output;
+import org.springframework.cloud.task.batch.listener.support.TaskBatchEventListenerBeanPostProcessor;
+import org.springframework.cloud.task.listener.TaskLifecycleListener;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.integration.gateway.GatewayProxyFactoryBean;
+import org.springframework.messaging.MessageChannel;
+
+/**
+ * @author Michael Minella
+ * @author Glenn Renfro
+ */
+@Configuration
+@ConditionalOnBean(value = { Job.class, TaskLifecycleListener.class })
+@ConditionalOnProperty(prefix = "spring.cloud.task.batch.events", name = "enabled", havingValue = "true", matchIfMissing = true)
+public class BatchEventAutoConfiguration {
+
+	@Bean
+	@ConditionalOnMissingBean
+	public TaskBatchEventListenerBeanPostProcessor batchTaskExecutionListenerBeanPostProcessor() {
+		return new TaskBatchEventListenerBeanPostProcessor();
+	}
+
+	@Configuration
+	@EnableBinding(BatchEventsChannels.class)
+	@ConditionalOnMissingBean(name = "jobExecutionEventsListener")
+	public static class JobExecutionListenerConfiguration {
+
+		@Autowired
+		private BatchEventsChannels listenerChannels;
+
+		@Bean
+		@Lazy
+		public GatewayProxyFactoryBean jobExecutionEventsListener() {
+			GatewayProxyFactoryBean factoryBean =
+					new GatewayProxyFactoryBean(JobExecutionListener.class);
+
+			factoryBean.setDefaultRequestChannel(listenerChannels.jobExecutionEvents());
+
+			return factoryBean;
+		}
+
+		@Bean
+		public StepExecutionListener stepExecutionEventsListener() {
+			return new EventEmittingStepExecutionListener(listenerChannels.stepExecutionEvents());
+		}
+
+		@Bean
+		@Lazy
+		public GatewayProxyFactoryBean chunkEventsListener() {
+			GatewayProxyFactoryBean factoryBean =
+					new GatewayProxyFactoryBean(ChunkListener.class);
+
+			factoryBean.setDefaultRequestChannel(listenerChannels.chunkEvents());
+
+			return factoryBean;
+		}
+
+		@Bean
+		@Lazy
+		public GatewayProxyFactoryBean itemReadEventsListener() {
```

Comment:
We probably don't want to use the `GatewayProxy` method for the read listener since that would result in every item being sent over the wire.  A custom listener that only does something on error is probably a better approach (I realize this may be different from what I said yesterday ;)